render/naga-agal/src/builder.rs

```diff
@@ -910,33 +910,49 @@ impl<'a> NagaBuilder<'a> {
             }
         };
 
+        // TODO - ideally, Naga would be able to tell us this information.
+        let source_is_scalar = matches!(
+            self.func.expressions[expr],
+            Expression::Math {
+                fun: MathFunction::Dot,
+                ..
+            } | Expression::As { .. }
+        );
+
+        // A scalar write occurs when we have exactly one component in the dest write mask.
+        let num_dest_components = [Mask::X, Mask::Y, Mask::Z, Mask::W]
+            .into_iter()
+            .filter(|mask| dest.write_mask.contains(*mask))
+            .count();
+
         // Optimization - use a Store instead of writing individual fields
         // when we're writing to the entire output register.
         if dest.write_mask.is_all() {
+            // Splat a scalar to the vector size, if necessary.
+            let expr = if source_is_scalar && num_dest_components != 1 {
```

Comment:
This is inside the `dest.write_mask.is_all()`, so `num_dest_components` will always be 4.